js/web/lib/wasm/jsep/init.ts

```diff
@@ -113,6 +113,36 @@ class ComputeContextImpl implements ComputeContext {
     return this.backend.device.limits.maxComputeWorkgroupStorageSize;
   }
 
+  isSubgroupsSupported(): boolean {
+    return (
+      this.backend.device.features.has('subgroups' as GPUFeatureName) ||
+      this.backend.device.features.has('chromium-experimental-subgroups' as GPUFeatureName)
+    );
+  }
+
+  isSubgroupsF16Supported(): boolean {
```

Comment:
Is it possible an adapter has subgroups and chromium-experimental-subgroups (and also shader-f16), but not subgroups-f16? If so, what would be the expectation?

Reply:
For a very few Vulkan devices, yes it is possible that `subgroups` and `shader-f16` (and `chromium-experimental-subgroups` as well, which enforce least checking to be available) features are supported but not `subgroups-f16`. This case means that the Vulkan backend doesn't support using f16 for subgroups built-in function in the shader and could end up in backend error if doing so.

But currently the subgroups feature proposal is not completely settled down, in the future version it is possible that the `subgroups-f16` feature may get deleted and `subgroups` feature get stricter to ensure backend support using f16 types for subgroups if shader-f16 is supported.